core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,781 @@
+	else
+		*privdata = jr_priv;
+
+	return retstatus;
+}
+
+/*
+ * Job Ring Interrupt handler
+ *
+ * Inputs:
+ * handler  Interrupt Handler structure
+ *
+ * Returns:
+ * ITRR_HANDLED  Interrupt has been handled
+ */
+static enum itr_return caam_jr_irqhandler(struct itr_handler *handler)
+{
+	struct jr_privdata *jrpriv = handler->data;
+
+	JR_TRACE("Disable the interrupt");
+	itr_disable(jrpriv->it_handler.it);
+
+	sev();
+
+	return ITRR_HANDLED;
+}
+
+
+/**
+ * Dequeues all Jobs completed. Call the job context callback
+ * function. Function returns the bit maks of the completed job
+ * expected (\a waitJobIds parameter)
+ *
+ * Inputs:
+ * waitJobIds  Expected Jobs to be complete
+ *
+ * Returns:
+ * Bit mask of the jobs completed function of input parameter
+ */
+static uint32_t do_jr_dequeue(uint32_t waitJobIds)
+{
+	uint32_t retJobId = 0;
+
+	struct caller_info   *caller = NULL;
+	struct outring_entry *jr_out = NULL;
+	struct jr_jobctx     *jobctx = NULL;
+	uint32_t  exceptions = 0;
+	int       found = -1;
+	uint16_t  idx_jr = 0;
+	uint16_t  nbJobs_done = 0;
+	size_t    nbJobs_inv = 0;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->inlock);
+
+	nbJobs_done = hal_jr_get_nbJobDone(jr_privdata->baseaddr);
+
+	if (nbJobs_done == 0) {
+		cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+		return retJobId;
+	}
+
+	/* Ensure that output ring descriptor entries are not in cache */
+	if ((jr_privdata->outread_index + nbJobs_done) > jr_privdata->nbJobs) {
+		/* Invalidate all job buffer */
+		/* Some job completed are at before the current outread_index */
+		jr_out     = jr_privdata->outrings;
+		nbJobs_inv = jr_privdata->nbJobs;
+	} else {
+		/* Invalidate only the job index completed */
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+		nbJobs_inv = nbJobs_done;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, (void *)jr_out,
+			(sizeof(struct outring_entry) * nbJobs_inv));
+
+	do {
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+
+		/*
+		 * Lock the caller information array because enqueue is
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
+		for ((idx_jr = 0), (found = (-1));
+				(idx_jr < jr_privdata->nbJobs) && (found < 0);
+				idx_jr++) {
+			/*
+			 * Search for the caller information corresponding to
+			 * the completed JR.
+			 * Don't use the outread_index or inwrite_index because
+			 * completion can be out of order compared to input
+			 * buffer
+			 */
+			caller = &jr_privdata->callers[idx_jr];
+			if (caam_read_val(&jr_out->desc) == caller->pdesc) {
+				jobctx         = caller->jobctx;
+				jobctx->status = caam_read_val(&jr_out->status);
+
+				/* Update return Job Id mask */
+				if (caller->jobid & waitJobIds)
+					retJobId |= caller->jobid;
+
+				JR_TRACE("JR id=%d, context @0x%08"PRIxVA"",
+					caller->jobid, (vaddr_t)jobctx);
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc = 0;
+				caller->jobid = JR_JOB_FREE;
+				found = idx_jr;
```

Comment:
Why is this needed when `jobctx` is assigned as well?
Putting this for loop in a helper function which would return the found `jobctx` would simplify this function a bit.